### PR TITLE
Fix recursive stack overflow issue on pathological recursive regular expressions

### PR DESCRIPTION
**Pull request by zv**
This patch changes the signature of `could_be_empty_branch` to include an argument that keeps track of recursive calls, which is a far better than my previous commits which fix a few specific problematic compiled `OP`s.

### CODE REVIEW

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by garazdawi** [comment]
wouldn't it be better if you just submit a patch to update our pcre support to the latest (i.e. pcre 8.39) instead of just taking a couple of patches?

**Comment by zv** [comment]
@garazdawi Erlang's PCRE library retains modifications not present upstream. I also went through a lot of hard work to find these bugs -- if it was hard for me it should be hard for you!

**Comment by garazdawi** [comment]
Hello @zv,

It was not my intent to diminish the work that you have put into this. Thanks for that and with some more effort we should be able to bring this patch into Erlang/OTP. 

If the pcre bug has been found and fixed upstream (which this one has, http://vcs.pcre.org/pcre?view=revision&sortby=rev&revision=1542), the patch to fix the issue should be supplied by including a diff of the change vs version 8.33 (aka svn rev 1336) together with the pcre bug id if there is one. The diff should then be applied to the relevant files as is together with the modifications needed to adjust it to run within erts. An example patch that you can look at is: https://github.com/erlang/otp/commit/6138bb612cd23c0ca37faadbe3fc9944be275228#diff-59347161e18d98230b255e855ff51f84. The code in erts pcre should not diverge at all from the upstream pcre except for the modifications needed to integrate it with erts which are all clearly marked by the define ERLANG_INTEGRATION. This is all necessary for us to keep track of what we have taken from upstream.

I know that this is quite a lot of work, especially since the fix you want to bring into erts seems to depend on other pcre patches, which is why I suggested that you might want to spend that time looking into upgrading the pcre version in erts instead. We will however of course also accept patches to individual bugs if you follow the template I described above.

Also, in case you've missed it, the ERLANG_INTEGRATION is described in greater details in https://github.com/erlang/otp/blob/maint/erts/emulator/pcre/README.pcre_update.md.

Thinking a little more long term about pcre in Erlang, maybe it is time to think about moving the entire library out into a nif and running it without modification on a dirty scheduler? That way Erlang could dynamically link with pcre and we wouldn't have to deal with not having bugfixes that are available upstream.

**Comment by zv** [comment]
I *think* you mean to add the EXIM bug ID and patch number -- I've done this here in the relevant commit.

**Comment by garazdawi** [comment]
@zv The main problem that I see with loading pcre dynamically is Windows. I've found a couple of pre-built dll's for it, but all of them are of ancient pcre versions. So keep that in mind while investigating. 

Regarding this PR, the change that you reference with r1498 only changes about 5-10 lines of code, while your change includes code from r1348, r1359 ... r1498. I want a complete list of all the patches that you are taking from upstream together with their associated upstream revisions and EXIM bug ID if available.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by zv** [comment]
I'm sorry, you are absolutely right -- I had an incorrect patch number on hand -- This should be the correct patch.

[Here is the reference patch](http://vcs.pcre.org/pcre/code/trunk/pcre_compile.c?r1=1348&r2=1347)

Again, I've split this patch over *two* commits to make merging a PR easier, let me know if this needs to be merged.

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

File: erts/emulator/pcre/pcre_compile.c [hunk 1]
```diff
@@ -2390,25 +2388,47 @@ for (code = first_significant_code(code + PRIV(OP_lengths)[*code], TRUE);
...
+      {
+      const pcre_uchar *tcode;
+      for (tcode = cd->start_workspace; tcode < cd->hwm; tcode += LINK_SIZE)
+        if ((int)GET(tcode, 0) == (int)(code + 1 - cd->start_code)) return TRUE;
+      if (GET(scode, 1) == 0) return TRUE;    /* Unclosed */
+      }
 
-    /* Not a forward reference, test for completed backward reference */
+    /* If the reference is to a completed group, we need to detect whether this
+    is a recursive call, as otherwise there will be an infinite loop. If it is
```
Comment (garazdawi) [inline erts/emulator/pcre/pcre_compile.c]: this comment is different from the upstream patch, why?

File: erts/emulator/pcre/pcre_compile.c [hunk 2]
```diff
@@ -2390,25 +2388,47 @@ for (code = first_significant_code(code + PRIV(OP_lengths)[*code], TRUE);
...
 
-    for (scode = cd->start_workspace; scode < cd->hwm; scode += LINK_SIZE)
-      if ((int)GET(scode, 0) == (int)(code + 1 - cd->start_code)) return TRUE;
+    if (cd->start_workspace != NULL)
+      {
+      const pcre_uchar *tcode;
+      for (tcode = cd->start_workspace; tcode < cd->hwm; tcode += LINK_SIZE)
+        if ((int)GET(tcode, 0) == (int)(code + 1 - cd->start_code)) return TRUE;
+      if (GET(scode, 1) == 0) return TRUE;    /* Unclosed */
+      }
```
Comment (garazdawi) [inline erts/emulator/pcre/pcre_compile.c]: in the upstream patch there is an "else {" here, why has it been moved?

File: erts/emulator/pcre/pcre_compile.c [hunk 3]
```diff
@@ -2335,34 +2333,34 @@ for (;;)
...
-
-Arguments:
-  code        points to start of search
-  endcode     points to where to stop
-  utf         TRUE if in UTF-8 / UTF-16 / UTF-32 mode
-  cd          contains pointers to tables etc.
-
-Returns:      TRUE if what is matched could be empty
+/* This function is called to check for left recursive calls. We want to check
+   the current branch of the current pattern to see if it could match the empty
```
Comment (garazdawi) [inline erts/emulator/pcre/pcre_compile.c]: this comment isn't changed in the upstream patch, why is it changed here?

**Comment by garazdawi** [comment]
That's better. There are still some discrepancies in between the upstream patch and what you have submitted, I added comments on the ones that I found.

Also please include the contents of http://vcs.pcre.org/pcre/code/trunk/pcre_compile.c?r1=1348&r2=1347&view=patch in the pr.

Do you know if there are any tests that should be updated to catch this bug? The tests which are taken from upstream are here: https://github.com/erlang/otp/tree/maint/lib/stdlib/test/re_SUITE_data

**Comment by garazdawi** [comment]
Closing this due to inactivity, please open a new PR when/if you decide to get back to it.

**Pull request closed without merging**